src/vmm/src/guest_config/x86/mod.rs

```diff
@@ -0,0 +1,41 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+
+use crate::guest_config::Modifier;
+
+/// Composite type that holistically provides
+/// the location of a specific register being used
+/// in the context of a CPUID tree
+#[allow(unused)]
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct CpuidRegisterId {
+    leaf: u32,
+    subleaf: u32,
+    pointer: CpuidRegister,
+}
+
+/// CPUID register enumeration
+#[allow(unused)]
+#[allow(missing_docs)]
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub enum CpuidRegister {
+    Eax,
+    Ebx,
+    Ecx,
+    Edx,
+}
+
+/// Template providing bitmask modifier info
+/// for CPUID and MSRs.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct X86CpuTemplate {
+    cpuid_modifiers: HashMap<CpuidRegisterId, Modifier>,
+    /// Map of Modifiers for CPU configuration
+    /// Key: Register pointer
+    /// Value: Register's bitmask based modifier
+    msr_modifiers: HashMap<u64, Modifier>,
+}
```

Comment:
Could we simplify `cpuid_modifiers` to `HashMap<CpuidKey, CpuidKey>` where
```rust
struct CpuidKey { leaf: u32, subleaf: u32 }
struct CpuidEntry { eax: u32, ebx: u32, ecx: u32, edx: u32 }
```

Reply:
We shouldn't forget about the flags somehow (which are properties of subleaves).

Reply:
Made edit to include the flags.

Reply:
To apply the registers without this I will need to find all the elements in the hash map then iterate through them, then map on the register field. I think this helps reduce duplicating the leaf & subleaf key and bundles the data together as it is conceptually.

Reply:
A similar approach to what you propose has been taken in [PR/3511](https://github.com/firecracker-microvm/firecracker/pull/3511/) for resolving duplication. Iteration won't be necessary though as the keys for the map will be addresses provided by the template.